File: CS/Cars.UWP/MainViewModel.cs
using Cars.Model;
using DevExpress.Mvvm;
using DevExpress.UI.Xaml.Scheduler;
using System.Collections.Generic;
using System.Linq;

namespace Cars.UWP {
    public class MainViewModel : ViewModelBase {


        CarsContext db;

        public MainViewModel() {
            db = new CarsContext();
            Cars = db.Cars.ToList();
            this.SaveChangesCommand = new DelegateCommand<CarScheduling>(SaveChanges);
            this.AppointmentAddedCommand = new DelegateCommand<CarScheduling>(AppointmentAdded);
            this.AppointmentRemovingCommand = new DelegateCommand<IEnumerable<CarScheduling>>(AppointmentRemoving);
            this.AppointmentRemovedCommand = new DelegateCommand<IEnumerable<CarScheduling>>(AppointmentRemoved);
            this.AppointmentUpdatingCommand = new DelegateCommand<CarScheduling>(AppointmentUpdating);
            Appointments = db.CarSchedulings.ToList();
        }

        protected IEnumerable<Car> cars;
        public IEnumerable<Car> Cars
        {
            get { return this.cars; }
            set { this.SetProperty(ref this.cars, value, "Cars"); }
        }


        protected IEnumerable<CarScheduling> appointments;
        public IEnumerable<CarScheduling> Appointments
        {
            get { return this.appointments; }
            set { this.SetProperty(ref this.appointments, value, "Appointments"); }
        }

        public DelegateCommand<CarScheduling> SaveChangesCommand { get; private set; }

        public DelegateCommand<CarScheduling> AppointmentAddedCommand { get; private set; }

        public DelegateCommand<IEnumerable<CarScheduling>> AppointmentRemovingCommand { get; private set; }
        public DelegateCommand<IEnumerable<CarScheduling>> AppointmentRemovedCommand { get; private set; }

        public DelegateCommand<CarScheduling> AppointmentUpdatingCommand { get; private set; }

        protected bool ProcessOccurranceChange = false;
        protected string RecurrenceRule = string.Empty;

        public void AppointmentUpdating(CarScheduling appointment)
        {
            if(appointment.EventType == (int)AppointmentType.Occurrence)
            {
                ProcessOccurranceChange = true;
            }

            if(appointment.EventType == (int)AppointmentType.Pattern)
            {
                RecurrenceRule = appointment.RecurrenceRule;
            }
        }

        public void AppointmentAdded(CarScheduling appointment)
        {
            db.Add(appointment);
            db.SaveChanges();
            Refresh();
        }

        public void AppointmentRemoving(IEnumerable<CarScheduling> appointments)
        {

            foreach (var item in appointments)
            {
                if(item.EventType != (int)AppointmentType.Occurrence)
                {
                    db.Remove(item);
                    if(item.EventType == (int)AppointmentType.Pattern)
                    {
                        RemoveOccurrances(item.PatternId);
                    }
                }                          
            }
                
            db.SaveChanges();
        }

        public void AppointmentRemoved(IEnumerable<CarScheduling> appointments) {

            foreach(var item in appointments)
            {
                if(item != null && item.EventType == (int)AppointmentType.DeletedOccurrence)
                {
                    db.Add(item);
                }
            }

            db.SaveChanges();
            Refresh();
        }


        void RemoveOccurrances(string patternId) {
            var occurrances = db.CarSchedulings.Where(c => c.PatternId == patternId
                        && (c.EventType == (int)AppointmentType.ChangedOccurrence || c.EventType == (int)AppointmentType.DeletedOccurrence)).ToList();
            db.RemoveRange(occurrances);
        }

        public void SaveChanges(CarScheduling appointment)
        {
            if(appointment != null && ProcessOccurranceChange)
            {
                if(appointment.EventType == (int)AppointmentType.ChangedOccurrence)
                {
                    db.Add(appointment);
                }
                ProcessOccurranceChange = false;
            }

            if(appointment.EventType == (int)AppointmentType.Pattern && appointment.RecurrenceRule != RecurrenceRule)
            {
                RemoveOccurrances(appointment.PatternId);
                RecurrenceRule = string.Empty;
            }

            db.SaveChanges();
            Refresh();
        }


        protected void Refresh() {
            this.Appointments = db.CarSchedulings.ToList();
        }
    }
}

File: VB/Cars.UWP/MainViewModel.vb
Imports Cars.Model
Imports DevExpress.Mvvm
Imports DevExpress.UI.Xaml.Scheduler

Public Class MainViewModel
    Inherits ViewModelBase

    Private db As CarsContext

    Public Sub New()
        db = New CarsContext()
        _cars = db.Cars.ToList()
        Me.SaveChangesCommand = New DelegateCommand(Of CarScheduling)(AddressOf SaveChanges)
        Me.AppointmentAddedCommand = New DelegateCommand(Of CarScheduling)(AddressOf AppointmentAdded)
        Me.AppointmentRemovingCommand = New DelegateCommand(Of IEnumerable(Of CarScheduling))(AddressOf AppointmentRemoving)
        Me.AppointmentRemovedCommand = New DelegateCommand(Of IEnumerable(Of CarScheduling))(AddressOf AppointmentRemoved)
        Me.AppointmentUpdatingCommand = New DelegateCommand(Of CarScheduling)(AddressOf AppointmentUpdating)
        appointments = db.CarSchedulings.ToList()
    End Sub

    Protected _cars As IEnumerable(Of Car)

    Public Property Cars As IEnumerable(Of Car)
        Get
            Return Me._cars
        End Get
        Set(ByVal value As IEnumerable(Of Car))
            Me.SetProperty(Me._cars, value, "Cars")
        End Set
    End Property

    Protected _appointments As IEnumerable(Of CarScheduling)

    Public Property Appointments As IEnumerable(Of CarScheduling)
        Get
            Return Me._appointments
        End Get
        Set(ByVal value As IEnumerable(Of CarScheduling))
            Me.SetProperty(Me._appointments, value, "Appointments")
        End Set
    End Property

    Public Property SaveChangesCommand As DelegateCommand(Of CarScheduling)
    Public Property AppointmentAddedCommand As DelegateCommand(Of CarScheduling)
    Public Property AppointmentRemovingCommand As DelegateCommand(Of IEnumerable(Of CarScheduling))
    Public Property AppointmentRemovedCommand As DelegateCommand(Of IEnumerable(Of CarScheduling))
    Public Property AppointmentUpdatingCommand As DelegateCommand(Of CarScheduling)
    Protected ProcessOccurranceChange As Boolean = False
    Protected RecurrenceRule As String = String.Empty

    Public Sub AppointmentUpdating(ByVal appointment As CarScheduling)
        If appointment.EventType = CInt(AppointmentType.Occurrence) Then
            ProcessOccurranceChange = True
        End If

        If appointment.EventType = CInt(AppointmentType.Pattern) Then
            RecurrenceRule = appointment.RecurrenceRule
        End If
    End Sub

    Public Sub AppointmentAdded(ByVal appointment As CarScheduling)
        db.Add(appointment)
        db.SaveChanges()
        Refresh()
    End Sub

    Public Sub AppointmentRemoving(ByVal appointments As IEnumerable(Of CarScheduling))
        For Each item In appointments

            If item.EventType <> CInt(AppointmentType.Occurrence) Then
                db.Remove(item)

                If item.EventType = CInt(AppointmentType.Pattern) Then
                    RemoveOccurrances(item.PatternId)
                End If
            End If
        Next

        db.SaveChanges()
    End Sub

    Public Sub AppointmentRemoved(ByVal appointments As IEnumerable(Of CarScheduling))
        For Each item In appointments

            If item IsNot Nothing AndAlso item.EventType = CInt(AppointmentType.DeletedOccurrence) Then
                db.Add(item)
            End If
        Next

        db.SaveChanges()
        Refresh()
    End Sub

    Private Sub RemoveOccurrances(ByVal patternId As String)
        Dim occurrances = db.CarSchedulings.Where(Function(c) c.PatternId = patternId AndAlso (c.EventType = CInt(AppointmentType.ChangedOccurrence) OrElse c.EventType = CInt(AppointmentType.DeletedOccurrence))).ToList()
        db.RemoveRange(occurrances)
    End Sub

    Public Sub SaveChanges(ByVal appointment As CarScheduling)
        If appointment IsNot Nothing AndAlso ProcessOccurranceChange Then

            If appointment.EventType = CInt(AppointmentType.ChangedOccurrence) Then
                db.Add(appointment)
            End If

            ProcessOccurranceChange = False
        End If

        If appointment.EventType = CInt(AppointmentType.Pattern) AndAlso appointment.RecurrenceRule <> RecurrenceRule Then
            RemoveOccurrances(appointment.PatternId)
            RecurrenceRule = String.Empty
        End If

        db.SaveChanges()
        Refresh()
    End Sub

    Protected Sub Refresh()
        Me.appointments = db.CarSchedulings.ToList()
    End Sub
End Class
